dissect/target/plugin.py

```diff
@@ -540,74 +531,160 @@ def decorator(obj):
+    DefaultPlugin, then plugins are returned as if no ``osfilter`` was
+    specified.
 
-    def _walk(osfilter: str = None, root: dict = None) -> Iterator[PluginDescriptor]:
-        for key, obj in root.items():
-            if key.startswith("_"):
-                continue
+    Another exeption to this are plugins in the ``PLUGINS`` tree which are
+    under a key that starts with a '_'. Those are only returned if their exact
+    key is specified in ``special_keys``.
 
-            if "functions" not in obj:
-                for plugin_desc in _walk(osfilter, obj):
-                    yield plugin_desc
-            else:
-                if osfilter and obj["module"].startswith("os") and not obj["module"].startswith(osfilter):
-                    continue
+    An exception to these exceptions is in the case of ``OSPlugin`` (sub-)class
+    plugins and ``os_filter`` is not ``None``. These plugins live in the
+    ``PLUGINS`` tree under the ``_os`` special key. Those plugins are only
+    returned if they fully match the provided ``osfilter``.
 
-                yield obj
+    The ``only_special_keys`` option returns only the plugins which are under a
+    special key that is defined in ``special_keys``. All filtering here will
+    happen as stated in the bove cases.
 
-    if (
-        osfilter
-        and isinstance(osfilter, type)
-        and issubclass(osfilter, OSPlugin)
-        and not issubclass(osfilter, general.default.DefaultPlugin)
-        and osfilter.__module__.startswith(MODULE_PATH)
-    ):
-        osfilter, _, _ = osfilter.__module__.replace(MODULE_PATH, "", 1).strip(".").rpartition(".")
-
-        # Continue walking up the OS filter tree until we hit the second level
-        # As an example, it walk os.unix.debian, followed by os.unix, then exit
-        os_parts = osfilter.split(".")
-        while len(os_parts) >= 2:
-            yield from _walk(".".join(os_parts), _get_plugins())
-            os_parts.pop()
-    else:
-        yield from _walk(None, _get_plugins())
+    Args:
+        osfilter: The optional OSPlugin to filter the returned plugins on.
+        special_keys: Also return plugins which are under the special ('_') keys in this set.
+        only_special_keys: Only return the plugins under the keys in ``special_keys`` and no others.
 
+    Yields:
+        Plugins in the ``PLUGINS`` tree based on the given filter criteria.
+    """
 
-def _special_plugins(special_key: str) -> Iterator[PluginDescriptor]:
-    """Retrieve plugin descriptors stored under ``special_key``."""
+    if osfilter is not None:
+        # The PLUGINS tree does not include the hierarchy up to the plugins
+        # directory (dissect.target.plugins) for the built-in plugins. For the
+        # plugins in the directory specified in --plugin-path, the hierarchy
+        # starts at that directory.
+        #
+        # Build-in OSPlugins do have the dissect.target.plugins path in their
```

Comment:
```suggestion
        # Built-in OSPlugins do have the dissect.target.plugins path in their
```

https://www.oxfordlearnersdictionaries.com/definition/english/built-in